start_control_ssm.sh

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+log() {
+  echo "$*" >&2
+}
+
+enable_hybrid_env_ssm() {
+  # SSM parameters necessary to register with a hybrid activation
+  activation_code=$(fetch_from_json ".ssm.activation-code" "${USER_DATA}")
+  activation_id=$(fetch_from_json ".ssm.activation-id" "${USER_DATA}")
+  region=$(fetch_from_json ".ssm.region" "${USER_DATA}")
+
+  # Register with AWS Systems Manager (SSM)
+  amazon-ssm-agent -register -code "${activation_code}" -id "${activation_id}" -region "${region}"
+  if [[ "$?" -ne 0 ]]; then
+    last_ssm_agent_error="$(tail -n 1 "/var/log/amazon/ssm/errors.log")"
+    log "Failed to register with AWS Systems Manager (SSM): ${last_ssm_agent_error}"
+    exit 1
+  fi
+}
+
+# Fetch the values from json, and exit on failure (if any)
+fetch_from_json() {
+  local key="${1:?}"
+  local file="${2:?}"
+  local value
+  if ! value=$(jq -e -r '$key' "${file}") || [[ -z "${value}" ]]; then
+    log "Failed to fetch value for ${key} from ${file}"
+    exit 1
+  fi
+  echo "${value}"
+}
+
+# Create the default directory structure on the persistent storage location
+prepare_persistent_state_dir() {
+  (
+    set -e
+    mkdir -p "${SSM_AGENT_PERSISTENT_STATE_DIR}"
+    chmod 750 "${SSM_AGENT_PERSISTENT_STATE_DIR}"
+    rmdir "${SSM_AGENT_LOCAL_STATE_DIR}"
+    ln -snf "${SSM_AGENT_PERSISTENT_STATE_DIR}" "${SSM_AGENT_LOCAL_STATE_DIR}"
+  )
+  if [[ "$?" -ne 0 ]]; then
+    log "Failed to prepare persistent state directory for SSM Agent"
+    exit 1
+  fi
+}
+
+# If /.bottlerocket/host-containers/current/user-data exists and is not empty
+# and the symlinked /var/lib/amazon/ssm/registration file is not populated,
+# then check to see if the user-data file contains ssm at the top-level. If so,
+# attempt to manually register with SSM with a hybrid activation.
+declare -r PERSISTENT_STORAGE_BASE_DIR="/.bottlerocket/host-containers/${HOST_CONTAINER_NAME}"
+declare -r USER_DATA="${PERSISTENT_STORAGE_BASE_DIR}/user-data"
+declare -r SSM_AGENT_PERSISTENT_STATE_DIR="${PERSISTENT_STORAGE_BASE_DIR}/ssm"
+declare -r SSM_AGENT_LOCAL_STATE_DIR="/var/lib/amazon/ssm"
+
+prepare_persistent_state_dir
+
+if [[ -s "${USER_DATA}" ]] \
+&& [[ ! -s "${SSM_AGENT_LOCAL_STATE_DIR}/registration" ]] \
+&& jq --exit-status '.ssm' "${USER_DATA}" &>/dev/null ; then
```

Comment:
This is basically `fetch_from_json`; we'd probably want to fail here, too, if `.ssm` was empty inside.

I think it might be easier to use `set -e` for the whole script.  Then you could just use `fetch_from_json` here too.  I think then you could just remove the `set -e` in `prepare_persistent_state_dir`, make `fetch_from_json` return 1 instead of exit 1, and change the main call to `if ! amazon-ssm-agent -bla` instead of checking `$?` afterward, and it'd be fine?